`original_console_log_level_config` reads from `logger.startup_console_log_level` instead of `logger.console_log_level`, so the console log level is never restored after startup

_Found via ClickGap automated review. Please close or comment if this is incorrect or needs adjustment._

_Retrospective finding from a historical scan of [PR #95919](https://github.com/ClickHouse/ClickHouse/pull/95919) (merged 2026-02-06). Confirmed on current codebase — close with a note if already fixed._

### Describe what's wrong

After server startup completes, the console log level is 'restored' to the startup override value instead of the original `console_log_level`. If a user configures `console_log_level=critical` and `startup_console_log_level=trace`, the console permanently stays at trace level instead of reverting to critical.

**Root cause:** Server.cpp:1227: Copy-paste error reads `logger.startup_console_log_level` instead of `logger.console_log_level`. Compare with the correct pattern at line 1210: `server_settings[ServerSetting::logger_level]` which correctly reads the DEFAULT level.

**Why we believe this is a bug:** Server.cpp:1227 → `config().getString("logger.startup_console_log_level", "")` saves the startup override value into `original_console_log_level_config`. At line 3062, this same (wrong) value is written back to `logger.console_log_level`, making the restore a no-op.

**Affected locations:**
- `programs/server/Server.cpp:1227` — original_console_log_level_config saves wrong config key value
- `programs/server/Server.cpp:3062` — restore writes back the startup value instead of the original

**Impact:** Users who set both `console_log_level` and `startup_console_log_level` will have their console permanently stuck at the startup override level after startup completes. This causes excessive console logging during normal server operation when `startup_console_log_level` is more verbose than `console_log_level`.

### Does it reproduce on most recent release?

Yes — confirmed on current `master` (commit `f17036af6b3a`).

### How to reproduce

```python
cd tests/integration && PATH=/path/to/clickhouse/build/programs:$PATH python3 -m pytest test_console_log_level_restore/test.py -v --timeout=180
```

### Expected behavior

```
PASSED - 'Restored console logger level to critical' should be found in the log
```

### Error message and/or stacktrace

```
FAILED - AssertionError: BUG: Console log level was restored to the startup_console_log_level value instead of the original console_log_level. Found 'Restored console logger level to trace': True. Expected 'Restored console logger level to critical': False.
```

### Additional context

**Open risks:**
- Documentation at _server_settings_outside_source.md uses wrong key names: `console_startup_log_level` and `console_shutdown_log_level` instead of the actual `startup_console_log_level` and `shutdown_console_log_level`

**Suggested fix:** Change line 1227 from `config().getString("logger.startup_console_log_level", "")` to `config().getString("logger.console_log_level", "")` to correctly save the original console log level before overriding it.

**Analysis details:** Confidence HIGH | Severity P2 | Testability: `INTEGRATION_TEST`

Found during automated review of [PR #95919](https://github.com/ClickHouse/ClickHouse/pull/95919).

---
_ClickGapAI · Confidence: HIGH · Severity: P2 · Finding: `h_pr95919_001`_


<!-- ch-version-info:start -->
### Version info
- Resolved by: #109858
- Merged into: `26.7.1.743` (included in `26.7` and later)
- Backported to: `26.6.2.63`, `26.5.6.35`, `26.4.5.124`, `26.3.17.40`
<!-- ch-version-info:end -->

--- a/programs/server/Server.cpp
+++ b/programs/server/Server.cpp
@@ -1342,7 +1342,8 @@ try
 
     // If the startup_console_log_level is set in the config, we override the console logger level.
     // Specific loggers can still override it.
-    std::string original_console_log_level_config = config().getString("logger.startup_console_log_level", "");
+    bool console_log_level_was_set = config().has("logger.console_log_level");
+    std::string original_console_log_level_config = config().getString("logger.console_log_level", "");
     bool should_restore_console_log_level = false;
     if (config().has("logger.startup_console_log_level") && !config().getString("logger.startup_console_log_level").empty())
     {
@@ -3447,9 +3448,20 @@ try
 
             if (should_restore_console_log_level)
             {
-                config().setString("logger.console_log_level", original_console_log_level_config);
-                Loggers::updateLevels(config(), logger());
-                LOG_INFO(log, "Restored console logger level to {}", original_console_log_level_config);
+                /// If the level was unset just remove the override so the default can be set via
+                /// Loggers::updateLevels again; otherwise restore the configured value.
+                if (console_log_level_was_set)
+                {
+                    config().setString("logger.console_log_level", original_console_log_level_config);
+                    Loggers::updateLevels(config(), logger());
+                    LOG_INFO(log, "Restored console logger level to {}", original_console_log_level_config);
+                }
+                else
+                {
+                    config().remove("logger.console_log_level");
+                    Loggers::updateLevels(config(), logger());
+                    LOG_INFO(log, "Restored console logger level to logger.level");
+                }
             }
 
             for (auto & server : servers)

--- a/tests/integration/test_server_startup_and_shutdown_logs/configs/config.d/logger_console_set.xml
+++ b/tests/integration/test_server_startup_and_shutdown_logs/configs/config.d/logger_console_set.xml
@@ -1,0 +1,7 @@
+<clickhouse>
+    <logger>
+        <level>information</level>
+        <console_log_level>warning</console_log_level>
+        <startup_console_log_level>trace</startup_console_log_level>
+    </logger>
+</clickhouse>

--- a/tests/integration/test_server_startup_and_shutdown_logs/configs/config.d/logger_console_unset.xml
+++ b/tests/integration/test_server_startup_and_shutdown_logs/configs/config.d/logger_console_unset.xml
@@ -1,0 +1,6 @@
+<clickhouse>
+    <logger>
+        <level>information</level>
+        <startup_console_log_level>trace</startup_console_log_level>
+    </logger>
+</clickhouse>

--- a/tests/integration/test_server_startup_and_shutdown_logs/test.py
+++ b/tests/integration/test_server_startup_and_shutdown_logs/test.py
@@ -12,6 +12,24 @@ instance = cluster.add_instance(
     stay_alive=True,
 )
 
+# Separate instances validate that the console log level is restored after startup.
+# They use a non-"none" root level so the post-startup "Restored ..." message is itself logged
+# (with level=none it would be suppressed right after the root logger level is restored).
+instance_console_unset = cluster.add_instance(
+    "node_console_unset",
+    main_configs=[
+        "configs/config.d/logger_console_unset.xml",
+    ],
+    stay_alive=True,
+)
+instance_console_set = cluster.add_instance(
+    "node_console_set",
+    main_configs=[
+        "configs/config.d/logger_console_set.xml",
+    ],
+    stay_alive=True,
+)
+
 
 @pytest.fixture(scope="module")
 def started_cluster():
@@ -20,6 +38,22 @@ def started_cluster():
         # Validate on startup we increase the log level.
         assert instance.contains_in_log("Starting root logger in level trace")
         assert instance.contains_in_log("Starting console logger in level trace")
+
+        # The console log level is raised at startup and then restored afterwards.
+        # When console_log_level is unset it reverts to follow logger.level ...
+        assert instance_console_unset.contains_in_log(
+            "Starting console logger in level trace"
+        )
+        assert instance_console_unset.contains_in_log(
+            "Restored console logger level to logger.level"
+        )
+        # ... and when it is explicitly configured it is restored to that value.
+        assert instance_console_set.contains_in_log(
+            "Starting console logger in level trace"
+        )
+        assert instance_console_set.contains_in_log(
+            "Restored console logger level to warning"
+        )
         yield cluster
     finally:
         cluster.shutdown()

--- a/tests/integration/test_server_startup_and_shutdown_logs/test.py
+++ b/tests/integration/test_server_startup_and_shutdown_logs/test.py
@@ -13,8 +13,6 @@ instance = cluster.add_instance(
 )
 
 # Separate instances validate that the console log level is restored after startup.
-# They use a non-"none" root level so the post-startup "Restored ..." message is itself logged
-# (with level=none it would be suppressed right after the root logger level is restored).
 instance_console_unset = cluster.add_instance(
     "node_console_unset",
     main_configs=[

